nixos/modules/services/web-apps/pixelfed.nix

```diff
@@ -0,0 +1,366 @@
+        };
+
+        password = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          description = lib.mdDoc ''
+            The database user's password.
+          '';
+        };
+
+      };
+
+      appKey = mkOption {
+        type = types.str;
+        description = lib.mdDoc '' A random
+          32-character string to be used as an encryption key. No default value;
+          use php artisan key:generate in the dataDir to generate. '';
+      };
+
+      maxUploadSize = mkOption {
+        type =  types.ints.positive;
+        default = 8;
+        description = lib.mdDoc ''
+      Max upload size in megabytes.
+      '';
+      };
+
+      hostName = mkOption {
+        type = types.str;
+        description = lib.mdDoc ''
+        FQDN for the pixelfed instance.
+      '';
+      };
+
+      nginx.enableACME = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc ''
+        Whether or not to enable ACME and let's encrypt for the pixelfed vhost.
+      '';
+      };
+
+
+      poolConfig = mkOption {
+        type = with types; attrsOf (oneOf [ int str bool ]);
+        default = {
+          "pm" = "dynamic";
+          "php_admin_value[error_log]" = "stderr";
+          "php_admin_flag[log_errors]" = true;
+          "catch_workers_output" = true;
+          "pm.max_children" = "32";
+          "pm.start_servers" = "2";
+          "pm.min_spare_servers" = "2";
+          "pm.max_spare_servers" = "4";
+          "pm.max_requests" = "500";
+        };
+
+        description = lib.mdDoc ''
+        Options for pixelfed's PHPFPM pool.
+      '';
+      };
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/pixelfed";
+        description = lib.mdDoc ''
+        Home directory of the `pixelfed` user which holds
+        the application's state.
+        '';
+      };
+
+    };
+  };
+
+
+  config = mkIf cfg.enable {
+    users.users.pixelfed = {
+      isSystemUser = true;
+      group = "nginx";
+      home = cfg.dataDir;
+      createHome = true;
+    };
+
+    environment.systemPackages = [ artisan ];
+
+    services.phpfpm.pools.pixelfed = {
+      user = "pixelfed";
+      group = "nginx";
+
+      phpPackage = pkgs.php80;
+
+      phpOptions = ''
+        post_max_size = ${toString cfg.maxUploadSize}M
+        upload_max_filesize = ${toString cfg.maxUploadSize}M
+        max_execution_time = 600;
+      '';
+
+      settings = {
+        "user" = "${user}";
+        "group" = "${group}";
+        "listen.owner" = "nginx";
+        "listen.group" = "nginx";
+        "listen.mode" = "0660";
+        "catch_workers_output" = "yes";
+      } // cfg.poolConfig;
+
+    };
+
+    systemd.services.pixelfed-data-setup = {
+      description = "Setup dataDir for pixelfed and change permissions";
+      wantedBy = [ "multi-user.target" ];
+      path = with pkgs; [ bash ];
+
+      script = ''
```

Comment:
I feel like there is no need to do this manually as `StateDirectoryMode` exist if needed